servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,547 @@
+                                    ((FlatMapPublisherSubscriber<?, ?>) t).replenishDemandAndClearSignalsQueued();
+                                } else {
+                                    break;
+                                }
+                            } catch (Throwable cause) {
+                                delayedCause = catchUnexpected(delayedCause, cause);
+                            }
+                        }
+
+                        if (t instanceof TerminalNotification) {
+                            sendToTarget((TerminalNotification) t); // if this throws its OK as we have terminated
+                        } else {
+                            t = terminalNotification;
+                            if (t != null && t != complete()) { // don't wait for demand to process an error
+                                sendToTarget((TerminalNotification) t); // if this throws its OK as we have terminated
+                            }
+                        }
+                    } else {
+                        assert emittedCount < prevPendingDemand;
+                        pendingDemandUpdater.accumulateAndGet(this, prevPendingDemand - emittedCount,
+                                FlowControlUtils::addWithOverflowProtection);
+                    }
+                } finally {
+                    tryAcquire = !releaseLock(emittingLockUpdater, this);
+                }
+            }
+
+            if (delayedCause != null) {
+                throwException(delayedCause);
+            }
+        }
+
+        private void tryRequestMoreFromUpstream() {
+            if (--upstreamDemand == 0) { // heuristic to replenish demand when it is exhausted.
+                assert subscription != null;
+                upstreamDemand = source.maxConcurrency;
+                subscription.request(source.maxConcurrency);
+            }
+        }
+
+        private static boolean needsDemand(Object item) {
+            return item != MAPPED_SOURCE_COMPLETE &&
+                    !(item instanceof FlatMapPublisherSubscriber) && !(item instanceof TerminalNotification);
+        }
+
+        private boolean sendToTarget(Object item) {
+            if (targetTerminated) {
+                // No notifications past terminal/cancelled
+                return false;
+            } else if (item == MAPPED_SOURCE_COMPLETE) {
+                tryRequestMoreFromUpstream();
+                return false;
+            } else if (item instanceof TerminalNotification) {
+                // Load the terminal notification in case an error happened after an onComplete and we override the
+                // terminal value.
+                TerminalNotification terminalNotification = this.terminalNotification;
+                assert terminalNotification != null;
+                sendToTarget(terminalNotification);
+                return false;
+            } else if (item instanceof FlatMapPublisherSubscriber) {
```

Comment:
nit: you could remove all the `else` as in each block you return.

Reply:
I'll just keep them as it clarifies the clauses are expected to be mutually exclusive and reduces the number of lines.